libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,770 @@
+	assert(cheri_gettag(tramp));
+
+	return (tramp);
+}
+
+typedef ssize_t slot_idx_t;
+
+static struct {
+	struct tramp_data *data;
+	struct tramp_map_kv {
+		_Atomic(ptraddr_t) key;
+		_Atomic(slot_idx_t) index;
+	} *map;
+	_Atomic(slot_idx_t) size;
+	_Atomic(slot_idx_t) writers;
+	int exp;
+} tramp_table;
+
+static struct {
+	_Atomic(struct tramp_pg *) head;
+	atomic_flag lock;
+} tramp_pgs = {
+	.lock = ATOMIC_FLAG_INIT
+};
+
+static slot_idx_t
+tramp_table_max_load(int exp)
+{
+	/* MAX_LOAD is 37.5% of capacity. */
+	return (3 << (exp - 3));
+}
+
+static struct tramp_map_kv *
+tramp_map_new(int exp)
+{
+	struct tramp_map_kv *map;
+
+	map = xmalloc(sizeof(*map) << exp);
+	for (size_t i = 0; i < (1 << exp); ++i)
+		map[i] = (struct tramp_map_kv) {
+			.key = 0,
+			.index = -1
+		};
+	return (map);
+}
+
+static struct tramp_data *
+tramp_data_expand(int exp)
+{
+	struct tramp_data *data;
+	/*
+	 * The table only needs to be as large as the MAX_LOAD.
+	 */
+	data = realloc(tramp_table.data, sizeof(*data) * tramp_table_max_load(exp));
+	if (data == NULL)
+		rtld_fatal("realloc failed");
+	return (data);
+}
+
+/* Public domain. Taken from https://github.com/skeeto/hash-prospector */
+static uint32_t
+pointer_hash(uint64_t key)
+{
+	uint32_t x = key ^ (key >> 32);
+	x ^= x >> 16;
+	x *= 0x21f0aaadU;
+	x ^= x >> 15;
+	x *= 0xd35a2d97U;
+	x ^= x >> 15;
+	return (x);
+}
+
+static slot_idx_t
+nextSlot(slot_idx_t slot, uint32_t hash, int exp)
+{
+	uint32_t mask = (1 << exp) - 1;
+	uint32_t step = (hash >> (32 - exp)) | 1;
+	return ((slot + step) & mask);
+}
+
+static void
+resizeTable(int exp)
+{
+	struct tramp_map_kv *map;
+	struct tramp_data *data;
+	ptraddr_t key;
+	uint32_t hash;
+	slot_idx_t size, slot;
+
+	assert(0 < exp && exp < 32);
+
+	free(tramp_table.map);
+	data = tramp_data_expand(exp);
+	map = tramp_map_new(exp);
+	size = atomic_load_explicit(&tramp_table.size, memory_order_relaxed);
+
+	for (slot_idx_t idx = 0; idx < size; ++idx) {
+		key = (ptraddr_t)data[idx].target;
+		hash = pointer_hash(key);
+		slot = hash;
+		do
+			slot = nextSlot(slot, hash, exp);
+		while (atomic_load_explicit(&map[slot].key,
+		    memory_order_relaxed) != 0);
+		atomic_store_explicit(&map[slot].key, key,
+		    memory_order_relaxed);
+		atomic_store_explicit(&map[slot].index, idx,
+		    memory_order_relaxed);
+	}
+
+	tramp_table.data = data;
+	tramp_table.map = map;
+	tramp_table.exp = exp;
+}
+
+void
+tramp_hook(int event, void *target, const Obj_Entry *obj, const Elf_Sym *def,
+    void *link, void *rcsp)
+{
+	Elf64_Word sym_num;
+	const char *sym;
+	const char *callee;
+
+	compart_id_t caller_id;
+	const char *caller;
+
+	struct utrace_rtld ut = {};
+
+	sym_num = def == NULL ? 0 : def->st_name;
+	sym = def == NULL ? "<unknown>" : strtab_value(obj, def->st_name);
+	callee = comparts.data[obj->compart_id]->name;
+
+	if ((cheri_getperm(link) & CHERI_PERM_EXECUTIVE) == 0)
+		caller_id = ((uintptr_t *)
+		    cheri_setaddress(rcsp, cheri_gettop(rcsp)))[-2];
```

Comment:
You just blindly trust the caller's stack to be sensible, both in terms of this not crashing and in terms of this not being faked, given's just an integer?

Reply:
This code runs only when tracing compartment transitions and is not really meant to be bullet-proof in terms of security. I'll add a comment about this in the man page.